sklearn/linear_model/tests/test_logistic.py

```diff
@@ -1824,3 +1853,40 @@ def test_scores_attribute_layout_elasticnet():
 
             avg_score_lr = cross_val_score(lr, X, y, cv=cv).mean()
             assert avg_scores_lrcv[i, j] == pytest.approx(avg_score_lr)
+
+
+def test_illconditioned_lbfgs():
+    # check that lbfgs converges even with ill-conditioned X
+    X, y = make_classification(n_samples=100, n_features=60, random_state=0)
+    X[:, 1] += 10000
+    X[:, 0] *= 10000
+    lr_pre = LogisticRegression(random_state=0, precondition=True)
+    with pytest.warns(None) as record:
+        lr_pre.fit(X, y)
+    assert len(record) == 0
+    loss_pre = _logistic_loss(
+        np.hstack([lr_pre.coef_.ravel(), lr_pre.intercept_]),
+        X, 2 * y - 1, 1)
+
+    lr = LogisticRegression(random_state=0, precondition=False)
+    with pytest.warns(ConvergenceWarning):
+        lr.fit(X, y)
+    loss = _logistic_loss(np.hstack([lr.coef_.ravel(), lr.intercept_]),
+                          X, 2 * y - 1, 1)
+    assert loss_pre < loss
+
+
+def test_logistic_loss_preconditioning():
+    # check _logistic_loss and _logistic_loss_grad with preconditioning
```

Comment:
`_logistic_loss_grad` is actually not called in this test. Also it should be `_logistic_loss_and_grad` instead.

Reply:
But there is already `test_logistic_loss_and_grad` that should cover this case...

Actually no, I think it's important to also check the loss value returned by `_logistic_loss_and_grad` in this test even if the gradient itself is checked in `test_logistic_loss_and_grad`.

Reply:
yeah I think this was helpful in debugging. It ensures the loss calculation is invariant to whether we do preconditioning or not.

It doesn't check the ``offset`` implementation right now, though. Should I add it? It's a bit more involved and thinking about it made a knot in my brain for a moment. Which I feel probably means I should do it?